### ISSUE TITLE
Secondary Dimension 'Email' is not working

### ISSUE THREAD

**Issue by dineshbabunr** [padams/Open-Web-Analytics#469]
Hi,

i am using OWA one of our website, its a SPA developed using react app, i am able to view the results. By using Email address in secondary dimension i am trying to view the data, i have configured the email and it got stored in DB( request and Document tables), but not populating the results what should i missing? 

 window.OWATracker.setGlobalEventProperty('user_email',window.currentUser.username)

![image](https://user-images.githubusercontent.com/44352312/75689147-b64cfc00-5cc6-11ea-8c68-7166122725d6.png)

**Comment by Maaiins** [comment]
Looks like some more than Email address dimensions are not returning any results.

I found out that the column `inbound_visitor_id` in table `owa_request` is always zero, so the join wont return any results anyway.

Maybe a similar issue like #471  

@dineshbabunr Can you please take a look into your db and confirm my research?

This is the executed SQL:

```
SELECT
   document_via_document_id.page_title as pageTitle,
   document_via_document_id.url as pageUrl,
   visitor_via_.user_email as userEmail,
   COUNT(request.id) as pageViews 
FROM
   owa_request as request 
   JOIN
      owa_document AS document_via_document_id 
      ON request.document_id = document_via_document_id.id 
   JOIN
      owa_visitor AS visitor_via_ 
      ON request.inbound_visitor_id = visitor_via_.id 
WHERE
   request.yyyymmdd BETWEEN '20200224' AND '20200302' 
   AND request.site_id = 'XXX' 
GROUP BY
   document_via_document_id.page_title,
   document_via_document_id.url,
   visitor_via_.user_email 
ORDER BY
   pageViews DESC LIMIT 100
```

**Comment by Maaiins** [comment]
I´ve done some research.

The field `inbound_visitor_id` is not used anywhere, but the entity still have the foreign key set. So the API will respect the correlation and want to join this columns.
The fix will be, just comment out the foreign key in the entity.

But there is also an bug when the API returns an empty resultset.
The `OWA.datagrid.gridColumnOrder` in `owa.resulSetExplorer.js` normally contains all columns of the grid. The data is generated through the resultset, but when the resultset is empty it can not be filled.
This data will also be used to determine the old dimension and substitute them from query string. When `OWA.datagrid.gridColumnOrder` is empty this won´t work. So the api request will add the failed dimension automatically to the join.
In conclusion the api request won't return any data after a empty resultset, because of an broken join.

**Comment by padams** [comment]
Closing as the fix is now merged into master. Great job here tracking this down.

**Comment by Maaiins** [comment]
Was quite heavy to figure this part out...

**Comment by padams** [comment]
Yeah. the resultset creation is the most complicated part the framework.

**Comment by xuzhe0205** [comment]
@dineshbabunr @Maaiins @padams  Hi guys. I'm working on a SPA project using React. It's a long question.

I am facing a problem where on a specific webpage (let's call it Cloud, and in React, this will be the <Cloud> component containing everything, with the path "/cloud"), if I click on a tab on the side navigation bar, this Cloud page (the <Cloud> component) will get reloaded since I would be navigated to a new path (say "/newpage") by <Link> tag. So now the content presented on the Cloud page changes (but will still be contained in this <Cloud> component). **Here comes my problem**, the OWA is able to track all other navigation (login, landing page etc.) and show the relative stats on my OWA management console website (showing number of page views and visits etc.), **but not for my aforementioned navigation by my side nav bar**. And the weirdest thing was it worked if I refreshed the this page (content contained by <Cloud>, with the same path "/newpage"):

For the "success" case where the OWA console website was able to track and show the stats, I inspected that the OWA event type was **owa_event_type: base.page_request**.  However, for my "failed" case where stats (showing number of page views and visits etc.) weren't showing on the OWA console website, my clicking on the side nav bar tab (which triggered routing by <Link>) did get tracked by OWA. BUT, the OWA event type was: **owa_event_type: dom.click**, **even though the URL did get changed from "/cloud" to "/newpage".**  

So **I am wondering is there a way to simply let OWA to track the page URL change, and record the relative stats in to the OWA console website?**

Btw, I put the JS code for tracking provided in this github wiki directly to the "index.html" (I tried putting it in `<head>` or `<body` tag, both worked).

Thanks for your help in advance!

**Comment by padams** [comment]
OWA doesn't listen for URL changes are anything like that. You would have to hook into the url change somehow and write a handler that tells OWA to track a second event.

**Comment by xuzhe0205** [comment]
> OWA listens for "click events" and fires new events to track any dom elements clicked on.
> 
> OWA doesn't listen for URL changes are anything like that. You would have to hook into the url change somehow and write a handler that tells OWA to track a second event.

Thanks for your rapid reply! So say I do that, will the new "URL" that user goes into get recorded and presented on the OWA console website (say in the "Content" section with number of page views, visits etc. listed) ?

 For example, after I make the code do as you said, if user clicks on some tabs on my nav bar (e.g. able to navigate with <a href="/newpage" >), will info like these be presented on OWA console website: "Page Path": "/newpage", "Page Title": ...., "Page Type": ...., "Page Views": 1, "Page Visit": 1, "Unique Views": 1 ?

**Comment by padams** [comment]
Yes, so long as you explicitly pass the URL was an argument in the trackPageView cmd.

**Comment by xuzhe0205** [comment]
> Yes, so long as you explicitly pass the URL was an argument in the trackPageView cmd.

Hi! Thanks for the help! It worked when I put the code `window.owa_cmds.push(["trackPageView"]);` in my React component that handles routing, and things seemed working out fine. But I'm curious how would you pass the URL as an argument in the trackPageView cmd? 

In comparison, I also tried using the "traditional flavor", and it worked with the URL passed as an argument like this: `OWATracker2.trackPageView(location.pathname);`

**Comment by padams** [comment]
`window.owa_cmds.push(["trackPageView", "argument1"]);`
